### PR TITLE
 Support build with Tensorflow

### PR DESCRIPTION
**Pull request by stweil**
These commits fix building of Tesseract with Tensorflow.

By default, `configure` will check automatically whether one of the Tensorflow header files which are needed by the Tesseract code exists. If yes, the build will use Tensorflow.

The `configure` options `--with-tensorflow` or `--without-tensorflow` can be used to explicitly require or refuse Tensorflow.

On Debian, installation of the packages `libtensorflow-dev` and `libprotoc-dev` is required. The include path is currently hardcoded for those to `/usr/include/tensorflow`.

### CODE REVIEW

**Comment by stweil** [comment]
It looks like `LSTMTrainer::InitTensorFlowNetwork` is currently unused, so training with TensorFlow needs more code. Text recognition with existing `traineddata` files with type `NT_TENSORFLOW` might work.

**Comment by Shreeshrii** [comment]
>Text recognition with existing traineddata files with type NT_TENSORFLOW might work.

Are there any traineddata files of this type?

**Comment by stweil** [comment]
> Are there any traineddata files of this type?

I don't know. Maybe @theraysmith, @jbreiden or @DanBloomberg know whether there exist TensorFlow models for Tesseract.

I know that other OCR software like for example [Calamari](https://github.com/Calamari-OCR/) uses TensorFlow and hope that the models from that software can be converted to be used by Tesseract.

**Comment by zdenop** [comment]
thanks!

**Comment by theceday** [comment]
This is not tf but keras, it is not the whole thing, but should help you to do the stuff.
It is kind of messy, including some constants in there and needs some refactoring, but nevertheless i think this is similar what tesseract does.
(at least thats what I think from looking at VGSL specs -same architecture but some different hyperparams-)

(this is for image_height=36)

making data generation and training completely two separate processes would be really good for people, allowing them to introduce their custom data.


`def customreshape(inputs):
    inner = inputs[0]
    shape = inputs[1]
    import tensorflow as tf2 
    reshaped = tf2.reshape(inner, [shape[0], shape[1], 54] )
    return reshaped

def tfshape(inputs):
    import tensorflow as tf2
    return tf2.shape(inputs)

def tfreshape(inputs):
    import tensorflow as tf2
    return tf2.reshape(inputs, [-1,18,16])

def build_model(self):
		input_shape = (self.img_h, None, 1)
        input_data = Input(name='inputs', shape=input_shape, dtype='float32')
        
        inner = Conv2D(self.conv_filters, self.kernel_size, strides=(1,1), padding='same', activation='relu', kernel_initializer='he_normal', name='convlayer')(input_data)
        inner = MaxPooling2D(pool_size=self.pool_size)(inner)
        
        inner = Permute((2, 1, 3)) (inner)        
        
        shape = Lambda(lambda x: tfshape(x), output_shape=(4,))(inner)        
        inner = Lambda(lambda x: tfreshape(x), output_shape=(18, 16))(inner)


        inner, h_, c_ = CuDNNLSTM(units=54, input_shape=(None, 16), return_sequences=False, return_state=True)(inner)

        
        inner = Lambda(lambda x : customreshape(x), output_shape=(None, 54))([inner, shape])

        inner = CuDNNLSTM(units=60, return_sequences=True, return_state=False, input_shape=(None, 54))(inner)
        
        inner = Bidirectional(CuDNNLSTM(units= self.rnn_size, input_shape=(None, 60), return_sequences=True))(inner)

        inner = Dense(self.alphabet_size +1 , kernel_initializer='he_normal')(inner)
        y_pred = Activation('softmax', name='softmax')(inner)
        Model(inputs=input_data, outputs=y_pred)

        labels = Input(name='the_labels', shape=[self.maxImageWidth // self.pool_size[1]], dtype='float32')
        input_length = Input(name='input_length', shape=[1], dtype='int64')
        label_length = Input(name='label_length', shape=[1], dtype='int64')

        loss_out = Lambda(ctc_lambda_func, output_shape=(1,), name='ctc')([y_pred, labels, input_length, label_length])
		..
		..

m = OCRModel(mode ='train', img_h=IMG_H, conv_filters=16, 
			 kernel_size=(3,3), pool_size= (2,2), rnn_size = 128, alphabetDict=dataProvider.alphabetDict,
			 absolute_max_string_len = dataProvider.maxYLen, maxImageWidth = MAXIMAGEWIDTH)
	
m.build_model()`
